demos/mac/cmac-aes256.c

```diff
@@ -0,0 +1,155 @@
+    0x78, 0xd4, 0x6d, 0x63, 0xf2, 0x2b, 0xe9, 0x56,
+};
+
+/*
+ * A property query used for selecting the MAC implementation.
+ */
+static const char *propq = NULL;
+
+int main(void)
+{
+    int rv = EXIT_FAILURE;
+    OSSL_LIB_CTX *library_context = NULL;
+    EVP_MAC *mac = NULL;
+    EVP_MAC_CTX *mctx = NULL;
+    unsigned char *out = NULL;
+    size_t out_len = 0;
+    OSSL_PARAM params[4], *p = params;
+    char cipher_name[] = "aes256";
+
+    library_context = OSSL_LIB_CTX_new();
+    if (library_context == NULL) {
+        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+        goto end;
+    }
+    OSSL_LIB_CTX_set0_default(library_context);
+
+    /* Fetch the CMAC implementation */
+    mac = EVP_MAC_fetch(library_context, "CMAC", propq);
+    if (mac == NULL) {
+        fprintf(stderr, "EVP_MAC_fetch() returned NULL\n");
+        goto end;
+    }
+
+    /* Create a context for the CMAC operation */
+    mctx = EVP_MAC_CTX_new(mac);
+    if (mctx == NULL) {
+        fprintf(stderr, "EVP_MAC_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /* The underlying cipher to be used */
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_MAC_PARAM_CIPHER, cipher_name,
+                                            sizeof(cipher_name));
+    *p = OSSL_PARAM_construct_end();
+
+    /* Initialise the CMAC operation */
+    if (!EVP_MAC_init(mctx, key, sizeof(key), params)) {
+        fprintf(stderr, "EVP_MAC_init() failed\n");
+        goto end;
+    }
+
+    /* Make one or more calls to process the data to be authenticated */
+    if (!EVP_MAC_update(mctx, data, sizeof(data))) {
+        fprintf(stderr, "EVP_MAC_update() failed\n");
+        goto end;
+    }
+
+    /* Make a call to the final with a NULL buffer to get the length of the MAC */
+    if (!EVP_MAC_final(mctx, NULL, &out_len, 0)) {
+        fprintf(stderr, "EVP_MAC_final() failed\n");
+        goto end;
+    }
+    out = OPENSSL_malloc(out_len);
+    if (out == NULL) {
+        fprintf(stderr, "malloc failed\n");
+        goto end;
+    }
+    /* Make one call to the final to get the MAC */
+    if (!EVP_MAC_final(mctx, out, &out_len, out_len)) {
+        fprintf(stderr, "EVP_MAC_final() failed\n");
+        goto end;
+    }
+
+    printf("Generated MAC:\n");
+    BIO_dump_indent_fp(stdout, out, out_len, 2);
+    putchar('\n');
+
+    if (out_len != sizeof(expected_output)) {
+        fprintf(stderr, "Generated MAC has an unexpected length\n");
+        goto end;
+    }
+
+    if (CRYPTO_memcmp(expected_output, out, sizeof(expected_output)) != 0) {
+        fprintf(stderr, "Generated MAC does not match expected value\n");
+        goto end;
+    }
+
+    rv = EXIT_SUCCESS;
+end:
+    if (rv != EXIT_SUCCESS)
+        ERR_print_errors_fp(stderr);
+    /* OpenSSL free functions will ignore NULL arguments */
+    OPENSSL_free(out);
+    EVP_MAC_CTX_free(mctx);
+    EVP_MAC_free(mac);
+    OSSL_LIB_CTX_free(library_context);
```

Comment:
Use valgrind.. This leaks currently. OSSL_LIB_CTX_set0_default() is probably the reason.